File: server/chat/prompt_builder.py
from typing import List, Dict
from server.prompts.chat import CHAT_SYSTEM_PROMPT


class PromptBuilder:
    """
    Phase 1.2
    ----------
    Builds a chat prompt with injected RAG context.
    No planning, no execution, no file writes.
    """

    def build(
        self,
        user_message: str,
        chunks: List[Dict]
    ) -> List[Dict]:
        messages = [
            {"role": "system", "content": CHAT_SYSTEM_PROMPT}
        ]

        if chunks:
            context_blocks = []
            for c in chunks:
                meta = c.get("metadata", {})
                context_blocks.append(
                    f"File: {meta.get('file_path')} "
                    f"(lines {meta.get('start_line')}–{meta.get('end_line')})\n"
                    f"```{meta.get('language', '')}\n"
                    f"{c.get('content')}\n```"
                )

            messages.append({
                "role": "system",
                "content": "CODE CONTEXT:\n\n" + "\n\n".join(context_blocks)
            })

        messages.append({"role": "user", "content": user_message})
        return messages

File: extension/src/features/chat/prompt-builder.ts
export class PromptBuilder {
  private static SYSTEM_PROMPT = (
    "You are a helpful AI assistant answering questions about a codebase.\n" +
    "You must base your answers ONLY on the provided context.\n" +
    "If the answer is not in the context, say \"I don't know\".\n" +
    "Do NOT suggest code changes or plans."
  );

  build(userMessage: string, chunks: any[]): Array<{ role: string; content: string }> {
    const messages: Array<{ role: string; content: string }> = [
      { role: "system", content: PromptBuilder.SYSTEM_PROMPT }
    ];

    if (chunks && chunks.length > 0) {
      const blocks = chunks.map((c: any) => {
        const m = c.metadata || {};
        const lang = m.language || "";
        const file = m.file_path || "";
        const start = m.start_line ?? "";
        const end = m.end_line ?? "";
        const content = c.content ?? "";
        return `File: ${file} (lines ${start}–${end})\n\u0060\u0060\u0060${lang}\n${content}\n\u0060\u0060\u0060`;
      });

      messages.push({
        role: "system",
        content: "CODE CONTEXT:\n\n" + blocks.join("\n\n")
      });
    }

    messages.push({ role: "user", content: userMessage });
    return messages;
  }
}
